dcpy/lifecycle/package/abstract_doc.py

```diff
@@ -0,0 +1,360 @@
+                    font=de.Font(bold=True, rgb=BLUE, size=TITLE_FONT_SIZE),
+                ),
+            ),
+        ],
+    )
+
+
+def _make_table_top(
+    title: str,
+    subtitle: str,
+    description: str | None = None,
+    maybe_image_path: Path | None = None,
+    column_headers: list[str] = [],
+    column_header_descriptions: list[str] = [],
+):
+    rows = []
+    if maybe_image_path:
+        rows.append(
+            de.Row(  # maybe the image cell
+                skip_default_styling=True,  # move
+                height=100,  # move
+                merge_cells=True,
+                cells=[de.Cell(value=de.Image(path=maybe_image_path))],
+            )
+        )
+
+    rows.append(
+        de.Row(  # subtitle cell
+            merge_cells=True,
+            is_top_row=True,  # move
+            height=30,  # move
+            cells=[_make_title_subtitle_cell(title, subtitle)],
+        )
+    )
+    if description:
+        rows.append(
+            de.Row(  # maybe the description cell
+                merge_cells=True,
+                height=50,
+                cells=[
+                    de.Cell(
+                        value=description,
+                        style=de.CellStyle(
+                            text_alignment_vertical="center",
+                            font=de.Font(italic=True, rgb=BLUE),
+                        ),
+                    ),
+                ],
+            )
+        )
+    if column_headers:
+        rows.append(
+            de.Row(
+                cells=[
+                    de.Cell(
+                        style=de.CellStyle(
+                            text_alignment_horizontal="center",
+                            text_alignment_vertical="center",
+                            font=de.Font(bold=True),
+                        ),
+                        value=col_name,
+                    )
+                    for col_name in column_headers
+                ]
+            )
+        )
+    if column_header_descriptions:
+        rows.append(
+            de.Row(
+                cells=[
+                    de.Cell(
+                        style=de.CellStyle(
+                            text_alignment_vertical="center",
+                            font=de.Font(italic=True, rgb=BLUE),
+                        ),
+                        value=desc,
+                    )
+                    for desc in column_header_descriptions
+                ]
+            ),
+        )
+
+    return rows
+
+
+def make_object_table(
+    *,
+    title: str,
+    subtitle: str,
+    table_rows: list[dict[str, str]],
+    description: str | None = None,
+    maybe_image_path: Path | None = None,
+    column_widths: list[float] | None = None,
+) -> de.Table:
+    """Make a table for an object.
+    Each row is an attribute of the object.
+    There are only two columns
+        - formatted key (summary + description)
+        - value
+    """
+    rows = _make_table_top(
```

Comment:
Not 100% on this - but at this level of abstraction it still feels slightly odd to me to have both the "top" and actual "table" all present as "row"s. Not that it's entirely improper - even for a non-excel output, to some extent it makes sense to turn each "table"/section(/tab) just into a series of rows/elements to be formatted. But it also seems similarly valid that these abstract "table"s could still have titles (or "header section", or whatever you want to call it - distinct entity from the tabular report itself), which could still already have some formatting but the individual formatters have their own opinion on how to put those together with the table.

That said, there's also some elegance in the specific formatters just getting a bunch of "row"s and knowing exactly what to do with each row. So really I'm just thinking out loud here without any helpful specific feedback lol

Reply:
But I think a good path forward, as said in person, is to test this out on html/pdfs and see what feels useful there